msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationKotlinTest.kt

```diff
@@ -101,6 +107,46 @@ class NativeAuthPublicClientApplicationKotlinTest : PublicClientApplicationAbstr
     private val password = "verySafePassword".toCharArray()
     private val code = "1234"
     private val emptyString = ""
+    private val allowPII = true
+    private val allowPIITrueToCheck = listOf(
+        """(?<![\[\(])["]password["][:=]?(?![\]\)])""", // '"password":' '"password"=' exclude 'password' '"challengeType":["password"]'
+        """(?<![\s\?\(])(code)[:=]""", // 'code:' 'code=' exclude 'codeLength' 'error?code',
+        """(?<![\(])continuationToken[:=]""",
+        """(?<![\(])attributes[:=]""",
+        """(?i)\b(accessToken|access_token)[:=]""", // access_token, accessToken
+        """(?i)\b(refreshToken|refresh_token)[:=]""",
+        """(?i)\b(idToken|id_token)[:=]""",
+        """(?i)\b(continuation_token)[:=]"""
+    )
+    private val allowPIIFalseToCheck = listOf(
+        """(?<![\(])username[:=]""",
+        """(?i)\b(challengeTargetLabel|challenge_target_label)[:=]""",
+        """(?i)\b(grantType|grant_type)[:=]"""
+    )
+    private val loggerCallback = object : ILoggerCallback {  // User anonymous object instead of lambda to avoid mockito VM not support issues
+        var failCalled = false
+
+        override fun log(
+            tag: String?,
+            logLevel: Logger.LogLevel?,
+            message: String?,
+            containsPII: Boolean
+        ) {
+            if (containsPII) {
+                val elementsToCheck = allowPIITrueToCheck.toMutableList()
+                if (!allowPII) {
+                    elementsToCheck.addAll(allowPIIFalseToCheck)
+                }
+                elementsToCheck.forEach { regex ->
+                    if (RegexMatcher(regex).matches(message)) {
+                        failCalled = true
+                        fail("PII $regex found in log message: $message")
```

Comment:
This is a smart setup, but I'm 100% sure it's going to work. Some thoughts: 
- I see we can't rely on only `fail`, as this callback is executed on a thread that's different from the thread that the test is executed on. I see that's why you've introduced the `failCalled` check, that's smart.
- however, given the fact that `failCalled` is set to true on a different thread, I expect there might be race conditions where the test thread is looking for `spyLoggerCallback.failCalled` (to perform `assertFalse`), before the callback has finished and set the value to `true`. given that the default is `false`, that means the test "thinks" failCalled wasn't set. Whereas in reality it hasn't been set _yet_
- you can try this out by adding a delay to the logger callback (e.g. `Thread.sleep(5000)`). 
- this setup will cause flaky tests: sometimes they'll pass, sometimes they'll fail. They're not deterministic, due to the race condition.
- typically a good way to deal with these async processes in tests is through for example `CountDownLatch`. Take a look at `LoggerTest.java`. Another good way in general when dealing with callbacks vs. coroutines is `suspendCancellableCoroutine` (this might not work here, but is interesting for you to look into).

Reply:
Thanks a million for this valuable guidance. Have added CountDownLatch into the loggercallback. But I just make changes for scenario 1 for reviewing first.